rules/ast/ast.go

```diff
@@ -749,55 +764,230 @@ func labelsEqual(labels1, labels2 clientmodel.Metric) bool {
+	// Calculate vector-to-vector operation.
+	if node.lhs.Type() == VectorType && node.rhs.Type() == VectorType {
+		lhs := node.lhs.(VectorNode).Eval(timestamp)
 		rhs := node.rhs.(VectorNode).Eval(timestamp)
-		for _, rhsSample := range rhs {
-			value, keep := evalVectorBinop(node.opType, lhs, rhsSample.Value)
-			if keep {
-				rhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					rhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-				}
-				result = append(result, rhsSample)
+
+		return node.evalVectors(timestamp, lhs, rhs)
+	}
+
+	// Calculate vector-to-scalar operation.
+	var lhs Vector
+	var rhs clientmodel.SampleValue
+	swap := false
+
+	if node.lhs.Type() == ScalarType && node.rhs.Type() == VectorType {
+		lhs = node.rhs.(VectorNode).Eval(timestamp)
+		rhs = node.lhs.(ScalarNode).Eval(timestamp)
+		swap = true
+	} else {
+		lhs = node.lhs.(VectorNode).Eval(timestamp)
+		rhs = node.rhs.(ScalarNode).Eval(timestamp)
+	}
+
+	result := make(Vector, 0, len(lhs))
+
+	for _, lhsSample := range lhs {
+		lv, rv := lhsSample.Value, rhs
+		// lhs always contains the vector. If the original position was different
+		// swap for calculating the value.
+		if swap {
+			lv, rv = rv, lv
+		}
+		value, keep := evalVectorBinop(node.opType, lv, rv)
+		if keep {
+			lhsSample.Value = value
+			if node.opType.shouldDropMetric() {
+				lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
 			}
+			result = append(result, lhsSample)
 		}
-		return result
-	} else if node.lhs.Type() == VectorType && node.rhs.Type() == ScalarType {
-		lhs := node.lhs.(VectorNode).Eval(timestamp)
-		rhs := node.rhs.(ScalarNode).Eval(timestamp)
-		for _, lhsSample := range lhs {
-			value, keep := evalVectorBinop(node.opType, lhsSample.Value, rhs)
-			if keep {
-				lhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
+	}
+	return result
+}
+
+// evalVectors evaluates the binary operation for the given vectors.
+func (node *VectorArithExpr) evalVectors(timestamp clientmodel.Timestamp, lhs, rhs Vector) Vector {
+	result := make(Vector, 0, len(rhs))
+	var err error
+	// The control flow below handles one-to-one or many-to-one matching.
+	// For one-to-many, swap sidedness and account for the swap when calculating
+	// values.
+	if node.matchCardinality == MatchOneToMany {
+		lhs, rhs = rhs, lhs
+	}
+	// All samples from the rhs hashed by the matching label/values.
+	rm := make(map[uint64]*Sample)
+	// Maps the hash of the label values used for matching to the hash of the label
+	// values of the include labels (if any.) It is used to keep track of already
+	// inserted samples.
+	added := make(map[uint64]uint64)
+
+	// Add all rhs samples to a map so we can easily find matches later.
+	for _, rs := range rhs {
+		hash := node.hashForMetric(rs.Metric.Metric)
+		// The rhs is guaranteed to be the 'one' side. Having multiple samples
+		// with the same hash means that the matching is many-to-many
```

Comment:
comma after "many-to-many" (because the "which" doesn't add further specificity, but only meta information)